pkgs/by-name/ha/haredo/setup-hook.sh

```diff
@@ -0,0 +1,68 @@
+haredoBuildPhase() {
+    runHook preBuild
+
+    local buildTargets jobs
+    read -ra buildTargets <<<"${haredoBuildTargets-}"
+    echoCmd "haredo build targets" "${buildTargets[@]}"
+    if [[ ! -v enableParallelBuilding || -n "${enableParallelBuilding-}" ]]; then
+        jobs="${NIX_BUILD_CORES}"
+    fi
+    haredo ${jobs:+"-j${jobs}"} "${buildTargets[@]}"
+
+    runHook postBuild
+}
+
+haredoCheckPhase() {
+    runHook preCheck
+
+    local checkTargets jobs
+
+    if [[ -n "${haredoCheckTargets:-}" ]]; then
+        read -ra checkTargets <<<"${haredoCheckTargets}"
+    else
+        for dofile in "check.do" "test.do"; do
+            [[ -r "${dofile}" ]] && {
+                checkTargets=("${dofile%".do"}")
+                break
+            }
+        done
+    fi
+
+    if [[ -z "${checkTargets:-}" ]]; then
+        printf -- 'no tests found; doing nothing\n'
```

Comment:
```suggestion
        printf -- 'no tests found; doing nothing\n'
        exit 1
```
I am a bit undecided about this but I think this would be a good setting to avoid silently no longer executing checks and no one notices it.

Reply:
```suggestion
        >&2 printf -- 'no tests found; doing nothing\n'
```
At least

Reply:
I've added @pbsds suggestion, since it's being stated that nothing is being done.

Reply:
The problem is that when the bot creates an auto update for a package and upstream changed the tests enough so that they are no longer run, we would never notice and assume all tests ran and everything is fine.

Reply:
Is it possible to detect if checkPhase is unset? If so, then raise a warning. If set to true, throw an error. If false, don't check

Reply:
that way a package may introduce tests in a bump which the nixpkgs build environ by default will try to run

Reply:
> Is it possible to detect if doCheck is unset? If so, then raise a warning. If set to true, throw an error. If false, don't check

I think this can be achieved by adding the `exit 1` call when the user has set `useHaredoCheck` --- the hooks equivalent of `doCheck` --- and no `checkTargets` were found. With it, when the tests are explicitly enabled, the build will fail if no `checkTargets` are found; moreover, if the user passes a non existent test, then `haredo` will fail, stopping the build.

What do you both think?

Reply:
Thinking it over, i believe `exit 1` is actually be fine since `doCheck` in `stdenv.mkDerivation` defaults to false.

Reply:
something like this?

```bash
if [[ -z "${dontHaredoCheck-}" && -z "${checkPhase-}" ]]; then
    checkPhase="haredoCheckPhase"
fi
```

and 

```bash
if [[ -z "${checkTargets:-}" ]]; then
  printf -- 'haredoCheckPhase ERROR: no check targets were found' 1>&2
  exit 1
fi
```

Reply:
I think it a bit verbose to have to set both `useHaredoCheck` and `doCheck` for `haredoCheckPhase` to be used.

Reply:
I see. Your changes do make the hook like the ones from other build frameworks like [just](https://github.com/NixOS/nixpkgs/blob/5d42b84e1b919cd2d44600db5d857d5d06e8efc9/pkgs/development/tools/just/setup-hook.sh#L52) and [ninja](https://github.com/NixOS/nixpkgs/blob/5d42b84e1b919cd2d44600db5d857d5d06e8efc9/pkgs/development/tools/build-managers/ninja/setup-hook.sh#L80).

I've left the check against `checkPhase`, so that the user may override it, even when using the `haredo`'s setup hook.

Reply:
May this conversation be resolved and the commits squashed?